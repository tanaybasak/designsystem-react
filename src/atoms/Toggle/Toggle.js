import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';

const Toggle = ({
  small,
  labelOff,
  labelOn,
  onChange,
  className,
  toggled,
  ...restProps
}) => {
  const [checked, setChecked] = useState(toggled);

  useEffect(() => {
    setChecked(toggled);
  }, [toggled]);

  const keyDownOnToggle = e => {
    const key = e.which || e.keyCode;
    if (key === 39) {
      e.preventDefault();
      setChecked(true);
      onChange(true, event);
    } else if (key === 37) {
      e.preventDefault();
      setChecked(false);
      onChange(false, event);
    }
  };

  return (
    <span
      className={`${prefix}-toggle${small ? '-small' : ''} ${className || ''}`}
    >
      <input
        type="checkbox"
        onChange={event => {
          setChecked(!checked);
          onChange(!checked, event);
        }}
        onKeyDown={keyDownOnToggle}
        checked={checked}
        {...restProps}
      />
      <label className={`${prefix}-toggle-label`} htmlFor={restProps.id}>
        <span className={`${prefix}-switch`} />
      </label>
      {labelOff ? (
        <span className={`${prefix}-toggle-off`}>{labelOff}</span>
      ) : null}
      {labelOn ? (
        <span className={`${prefix}-toggle-on`}>{labelOn}</span>
      ) : null}
    </span>
  );
};

Toggle.propTypes = {
  /** Boolean value to change toggle to small */
  small: PropTypes.bool,
  /** Off label for Toggle */
  labelOff: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** On label for Toggle */
  labelOn: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Call back function that is invoked when Toggle is clicked
   *
   * @value : bool value of toggle
   * @event : event when toggled
   */
  onChange: PropTypes.func,
  /** Name of the custom class to apply to the Toggle */
  className: PropTypes.string,
  /** Unique id for Toggle */
  id: PropTypes.string.isRequired,
  /** Boolean value representing state of Toggle Component */
  toggled: PropTypes.bool,
  /** Boolean value to disable Toggle */
  disabled: PropTypes.bool
};

Toggle.defaultProps = {
  small: false,
  labelOff: 'Off',
  labelOn: 'On',
  onChange: () => {},
  disabled: false,
  toggled: false
};

export default Toggle;
