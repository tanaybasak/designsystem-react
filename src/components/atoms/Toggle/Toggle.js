import React , { useState }from 'react';
import PropTypes from 'prop-types';
import prefix from '../../../settings';

const Toggle = ({
  small,
  labelOff,
  labelOn,
  onChange,
  className,
  toggled,
  ...restProps
}) => {
  const [checked, setChecked] = useState(toggled || false);

  return (
    <span
      className={`${prefix}-toggle${small ? "-small" : ""} ${className || ""}`}
    >
      <input
        type='checkbox'
        onChange={event => {
          setChecked(!checked);
          onChange(event);
        }}
        checked={checked}
        {...restProps}
      />
      <label className={`${prefix}-toggle-label`} htmlFor={restProps.id}>
        <span className={`${prefix}-switch`} />
      </label>
      {labelOff ? (
        <span className={`${prefix}-toggle-off`}>{labelOff}</span>
      ) : null}
      {labelOn ? (
        <span className={`${prefix}-toggle-on`}>{labelOn}</span>
      ) : null}
    </span>
  );
}

Toggle.propTypes = {
  small: PropTypes.bool,
  labelOff: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  labelOn: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func,
  className: PropTypes.string,
  id: PropTypes.string.isRequired,
  toggled: PropTypes.bool,
  disabled: PropTypes.bool
};

Toggle.defaultProps = {
  small: false,
  labelOff: "Off",
  labelOn: "On",
  onChange: () => {},
  disabled: false
};

export default Toggle;
