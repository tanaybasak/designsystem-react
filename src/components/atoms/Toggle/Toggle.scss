/******************************************************************
 * HCL Confidential
 *
 * Copyright HCL Technologies Ltd. 2019 All rights reserved.
 *
 *******************************************************************
 */
 
 $prefix: 'hcl';

 .#{$prefix}-toggle-container {
    display: inline-block;

    .#{$prefix}-toggle {
        position: relative;
        cursor: pointer;
        display: inline-block;
        width: 3rem;
        height: 1.5rem;
        top: 0;
        margin-left: 0.5rem;
        margin-right: 0.5rem;
        margin-bottom: 0;

        &:before {
            content: "";
            position: relative;
            width: 3rem;
            display: block;
            top: 50%;
            left: 0;
            background-color: #717171; // TODO: Need to check with design team
            transition: .4s;
            transform: translate(0, -50%);
            height: 0.25rem;
        }

        .#{$prefix}-slider {
            position: absolute;
            left: 0;
            display: block;
            background: white;
            height: 1.5rem;
            width: 1.5rem;
            top: 50%;
            background-color: #F5F5F5; // TODO: Need to check with design team
            border: 0.125rem solid #717171; // TODO: Need to check with design team
            transition: .4s;
            transform: translate(0, -50%);
            border-radius: 50%;
        }
    }

    input[type="checkbox"] {
        opacity: 0;
        width: 0;
        height: 0;
        position: absolute;

        &:checked+.#{$prefix}-toggle {
            &:before {
                background-color: var(--blue);
            }

            .#{$prefix}-slider {
                top: 50%;
                transform: translate(1.5rem, -50%);
                background-color: var(--blue);
                border: none;
            }
        }

        &:focus+.#{$prefix}-toggle {
            .#{$prefix}-slider {
                box-shadow: 0 0 0 0.125rem #5AAAFA; // TODO: Need to check with design team
                border: 0.125rem solid #5A6872;
            }
        }

        &:hover+.#{$prefix}-toggle {
            .#{$prefix}-slider {
                border: 0.125rem solid #383838; // TODO: Need to check with design team
            }
        }

        &:disabled+.#{$prefix}-toggle {
            &:before {
                background-color: #D2D2D2; // TODO: Need to check with design team
                cursor: not-allowed;
            }

            .#{$prefix}-slider {
                background-color: #F5F5F5; // TODO: Need to check with design team
                border: 0.125rem solid #D2D2D2; // TODO: Need to check with design team
                cursor: not-allowed;
            }
        }

        &:disabled~.#{$prefix}-toggle-off,
        &:disabled~.#{$prefix}-toggle-on {
            color: #AAAAAA; // TODO: Need to check with design team
        }
    }

    .#{$prefix}-toggle-off,
    .#{$prefix}-toggle-on {
        font-size: .875rem;
        font-weight: 400;
        line-height: 1.5rem;
        letter-spacing: .01rem;
        color: #0F212E; // TODO: Need to check with design team
    }

    .#{$prefix}-toggle-off {
        float: left;
    }

    .#{$prefix}-toggle-on {
        float: right;
    }

    &:after {
        content: '';
        display: block;
        clear: both;
    }
}
