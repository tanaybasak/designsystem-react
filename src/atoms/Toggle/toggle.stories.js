import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, boolean } from '@storybook/addon-knobs';
//@update-path-build-start
import Toggle from './Toggle';
//@update-path-build-end

storiesOf('Toggle', module)
  .add(
    'default',
    () => (
      <Toggle
        disabled={boolean('Disabled', false)}
        id="toggle-element"
        labelOff={text('Label Off', 'Off')}
        labelOn={text('Label On', 'On')}
        onChange={action('Toggle-OnChange')}
      />
    ),
    {
      info: {
        text: `Description About Toggle Component

        import { Toggle } from '@patron/patron-react/toggle'

        `
      }
    }
  )
  .add(
    'small',
    () => (
      <Toggle
        disabled={boolean('Disabled', false)}
        id="toggle-element"
        labelOff={text('Label Off', 'Off')}
        labelOn={text('Label On', 'On')}
        onChange={action('Toggle-OnChange')}
        small
      />
    ),
    {
      info: {
        text: `Description About Toggle Component

        import { Toggle } from '@patron/patron-react/toggle'

        `
      }
    }
  );
