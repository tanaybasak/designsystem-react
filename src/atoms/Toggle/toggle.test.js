import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import prefix from '../../settings';
import Toggle from './Toggle';

describe('<Toggle> component', () => {
    it('snapshots/renders Toggle correctly', () => {
        const tree = renderer.create(<Toggle id="hellotoggle"/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    // it('renders checkobx with label', () => {
    //     const checkbox = mount((<Toggle labelText='defaultcheckbox' />));
    //     expect(checkbox.text()).toBe('defaultcheckbox');
    // });

    // it('renders checkobx with default classname', () => {
    //     const checkbox = mount((<Toggle />));
    //     expect(checkbox.find(`.${prefix}-checkbox`).hasClass('hcl-checkbox')).toBe(true);
    // });
});