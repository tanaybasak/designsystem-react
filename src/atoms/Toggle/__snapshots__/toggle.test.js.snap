// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle> component snapshots/renders Toggle correctly 1`] = `
<span
  className="hcl-toggle "
>
  <input
    checked={false}
    disabled={false}
    id="hellotoggle"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="hcl-toggle-label"
    htmlFor="hellotoggle"
  >
    <span
      className="hcl-switch"
    />
  </label>
  <span
    className="hcl-toggle-off"
  >
    Off
  </span>
  <span
    className="hcl-toggle-on"
  >
    On
  </span>
</span>
`;
